### ISSUE TITLE
Can't process EDB UTF-16 encoding error

### ISSUE THREAD

**Issue by Lenny124** [libyal/libesedb#54]
Hello,

I got the following error:

libuna_unicode_character_copy_from_utf16_stream: missing surrogate UTF-16 character bytes.
libuna_utf8_string_size_from_utf16_stream: unable to copy Unicode character from UTF-16 stream.
libfvalue_string_get_utf8_string_size: unable to determine UTF-8 string size of UTF-16 stream.
libuna_unicode_character_copy_from_utf16_stream: missing surrogate UTF-16 character bytes.
libuna_utf8_string_with_index_copy_from_utf16_stream: unable to copy Unicode character from UTF-16 stream.
libfvalue_string_copy_to_utf8_string_with_index: unable to copy UTF-16 stream to UTF-8 string.
Unable to export file.
libuna_unicode_character_copy_to_utf8: UTF-8 string too small.
libuna_utf8_string_with_index_copy_from_byte_stream: unable to copy Unicode character to UTF-8.
libfvalue_string_copy_to_utf8_string_with_index: unable to copy byte stream to UTF-8 string.
libfvalue_value_copy_to_utf8_string_with_index: unable to copy instance to UTF-8 string.
libfvalue_value_copy_to_utf8_string: unable to copy value: 0 to UTF-8 string.
libesedb_record_value_get_utf8_string: unable to copy value to UTF-8 string.
libesedb_long_value_get_utf8_string: unable to retrieve UTF-8 string from record value.
export_handle_export_long_record_value: unable to retrieve value string: 39 (269).
export_handle_export_record_value: unable to export long record value: 39.
export_handle_export_record: unable to export record value: 39.
export_handle_export_table: unable to export record.
export_handle_export_file: unable to export table: 230.

Processed with esedbexport 20200418 on ubuntu. I build source code from github - latest release.

**Comment by Lenny124** [comment]
Pls help

**Comment by joachimmetz** [comment]
Per [Read first](https://github.com/libyal/libesedb/wiki/Building#read-first). You only need to build from git source if you're developing on this project otherwise use the source distribution package.

**Comment by joachimmetz** [comment]
Can you provide me with config.log and [debug output](https://github.com/libyal/libesedb/wiki/Troubleshooting#verbose-and-debug-output)

**Comment by Lenny124** [comment]
Hi Joachim,
Here is my config file which comes from:

`CPPCFLAGS=-g ./configure --enable-shared=no`
 
Config log:

[config.log](https://github.com/libyal/libesedb/files/4715262/config.log)

I ran with gdb:

`gdb --ex r --args ./esedbexport Windows.edb`

but couldn't create backtrace beacuse of this error:

`Unable to export file.
libuna_unicode_character_copy_to_utf8: UTF-8 string too small.
libuna_utf8_string_with_index_copy_from_byte_stream: unable to copy Unicode character to UTF-8.
libfvalue_string_copy_to_utf8_string_with_index: unable to copy byte stream to UTF-8 string.
libfvalue_value_copy_to_utf8_string_with_index: unable to copy instance to UTF-8 string.
libfvalue_value_copy_to_utf8_string: unable to copy value: 0 to UTF-8 string.
libesedb_record_value_get_utf8_string: unable to copy value to UTF-8 string.
libesedb_long_value_get_utf8_string: unable to retrieve UTF-8 string from record value.
export_handle_export_long_record_value: unable to retrieve value string: 39 (269).
export_handle_export_record_value: unable to export long record value: 39.
export_handle_export_record: unable to export record value: 39.
export_handle_export_table: unable to export record.
export_handle_export_file: unable to export table: 230.
[Inferior 1 (process 25992) exited with code 01]
(gdb) bt
No stack.
(gdb) 
`

This time error seems to be UTF-8 encoding isssue. Weird. 

Previously I configured source code with the following command:

` ./configure --enable-debug-output --enable-verbose-output --enable-wide-character-type`

**Comment by joachimmetz** [comment]
Run esedbexport with `-v` to generate the debug output

**Comment by Lenny124** [comment]
I ran with `-v`

`Unable to export file.
libuna_unicode_character_copy_to_utf8: UTF-8 string too small.
libuna_utf8_string_with_index_copy_from_byte_stream: unable to copy Unicode character to UTF-8.
libfvalue_string_copy_to_utf8_string_with_index: unable to copy byte stream to UTF-8 string.
libfvalue_value_copy_to_utf8_string_with_index: unable to copy instance to UTF-8 string.
libfvalue_value_copy_to_utf8_string: unable to copy value: 0 to UTF-8 string.
libesedb_record_value_get_utf8_string: unable to copy value to UTF-8 string.
libesedb_long_value_get_utf8_string: unable to retrieve UTF-8 string from record value.
export_handle_export_long_record_value: unable to retrieve value string: 39 (269).
export_handle_export_record_value: unable to export long record value: 39.
export_handle_export_record: unable to export record value: 39.
export_handle_export_table: unable to export record.
export_handle_export_file: unable to export table: 230.
`

**Comment by Lenny124** [comment]
The output seems to be the same...

**Comment by joachimmetz** [comment]
> The output seems to be the same...

then debug output is not enabled

**Comment by Lenny124** [comment]
I run the command with:

`./esedbexport -v WIndows.edb`

What could be wrong?

**Comment by joachimmetz** [comment]
after running:
`./configure --enable-debug-output --enable-verbose-output --enable-wide-character-type`

did you rebuild the source? e.g. `make clean && make`

Also did you try running the source packaged version? Not the version directly from git HEAD?

**Comment by Lenny124** [comment]
Joachim,

I run latest release version. I don't use version from git HEAD.

I'm running with verbose output now. It is taking much more time to process. When the crash is generated I send you the output.

**Comment by joachimmetz** [comment]
> I run latest release version. I don't use version from git HEAD.

Thx for confirming.

> I'm running with verbose output now. It is taking much more time to process. 

Excellent, that is expected. I generates a lot of data. I would be interested first mainly in the last 100 or so lines of the output (if the output becomes large).

**Comment by Lenny124** [comment]
It's still processing. I found the issue occuring on more than one EDB. I will try to generate more crashes.

BTW I processed one EDB and it was OK - I mean no crash occurred. Is there any way to associate the attachment (from Attachment* tables) to specific user or his mailbox ?

**Comment by joachimmetz** [comment]
> I will try to generate more crashes.

FYI it's not a crash (as in https://en.wikipedia.org/wiki/Crash_(computing)). You are hitting a fail-safe since the format currently does not make sense to libesedb. Hopefully the debug output will indicate what format edge case you are running into.

**Comment by joachimmetz** [comment]
> Is there any way to associate the attachment (from Attachment* tables) to specific user or his mailbox ?

Not sure what you are asking? What EDB file are you looking at that of Windows (Desktop) Search?

**Comment by Lenny124** [comment]
I'm trying to parse Exchange EDB file. Your tool produces a lot of tables with suffixes like pi1_1234 table etc. Do they have any meaning? Do you have any documentation of data produced.

**Comment by joachimmetz** [comment]
Yes, that is MAPI database you are looking at. The format of the application specific Exchange MAPI is outside the scope of this project. This project only focuses on the underlying ESE database layer not the application specific tables and record values.
